pkg/util/push/otel.go

```diff
@@ -0,0 +1,150 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package push
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+
+	kitlog "github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/tenant"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/translator/prometheusremotewrite"
+	"github.com/prometheus/prometheus/prompb"
+	"github.com/weaveworks/common/middleware"
+	"go.opentelemetry.io/collector/model/otlpgrpc"
+	"go.opentelemetry.io/collector/model/pdata"
+
+	"github.com/grafana/mimir/pkg/mimirpb"
+	"github.com/grafana/mimir/pkg/util/log"
+	"github.com/grafana/mimir/pkg/util/validation"
+)
+
+const (
+	pbContentType   = "application/x-protobuf"
+	jsonContentType = "application/json"
+
+	otelParseError = "otlp_parse_error"
+	maxErrMsgLen   = 1024
+
+	messageSizeLargerErrFmt = "received message larger than max (%d vs %d)"
+)
+
+func HandlerForOTLP(
+	maxRecvMsgSize int,
+	sourceIPs *middleware.SourceIPExtractor,
+	allowSkipLabelNameValidation bool,
+	push Func,
+) http.Handler {
+	return handler(maxRecvMsgSize, sourceIPs, allowSkipLabelNameValidation, push, func(ctx context.Context, r *http.Request, maxSize int, dst []byte, req *mimirpb.PreallocWriteRequest) ([]byte, error) {
+		var decoderFunc func(buf []byte) (otlpgrpc.MetricsRequest, error)
+
+		logger := log.WithContext(ctx, log.Logger)
+
+		contentType := r.Header.Get("Content-Type")
+		switch contentType {
```

Comment:
Should this check the contentType prefix instead? This could be `application/json; charset=utf-8`.